server/data/data-postgres/src/main/java/org/apache/james/user/postgres/PostgresUsersDAO.java

```diff
@@ -141,4 +150,93 @@ public void addUser(Username username, String password) {
                 e -> new AlreadyExistInUsersRepositoryException("User with username " + username + " already exist!"))
             .block();
     }
+
+    public Mono<Void> addAuthorizedUser(Username baseUser, Username userWithAccess, boolean targetUserExists) {
+        return addUserToList(AUTHORIZED_USERS, baseUser, userWithAccess)
+            .then(addDelegatedUser(baseUser, userWithAccess, targetUserExists));
+    }
+
+    private Mono<Void> addDelegatedUser(Username baseUser, Username userWithAccess, boolean targetUserExists) {
+        if (targetUserExists) {
+            return addUserToList(DELEGATED_USERS, userWithAccess, baseUser);
+        } else {
+            return Mono.empty();
+        }
+    }
+
+    private Mono<Void> addUserToList(Field<String[]> field, Username baseUser, Username targetUser) {
+        String fullAuthorizedUsersColumnName = TABLE.getName() + "." + field.getName();
+        return postgresExecutor.executeVoid(dslContext ->
+            Mono.from(dslContext.insertInto(TABLE_NAME)
+                .set(USERNAME, baseUser.asString())
+                .set(field, DSL.array(targetUser.asString()))
+                .onConflict(USERNAME)
+                .doUpdate()
+                .set(DSL.field(field.getName()),
+                    (Object) DSL.field("array_append(coalesce(" + fullAuthorizedUsersColumnName + ", array[]::varchar[]), ?)",
```

Comment:
string concatenation in a SQL query on a field potentially controlled by a user spotted.

Extra care handling this is definitively needed.

Reply:
String fullAuthorizedUsersColumnName = TABLE.getName() + "." + field.getName();

TABLE is constant. field could be AUTHORIZED_USERS or DELEGATED_USERS which are all constant. Therefore, it is not vulnerable.

Reply:
I did miss read.

Yes of course that's ok.